### DISCUSSION TITLE
Titanium Alloy Textfield on Focus

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13564] [Q&A]
I want to add event listener "on focus" for the $.txtOne element. I tried different methods. 

1) Via Alloy with "onFocus=filter" property but then no function is executed.  The property "onChange" works as expected.
2) With Titanium by calling `addEventListener` function and by define my own `addEventListener` function. But both ways throw same error:
>  [ERROR] V8Exception: Exception occurred at /alloy/controllers/test.js: 100: Uncaught TypeError: $.txtOne.addEventListener is not a function

But don't understand why `addEventListener` is not a function, even if I defined my own function `controllers/components/textfield.js`? For example the definition of `$.txtOne.getInputText()` is working. 

views/components/textfield.xml

```
<Alloy>
  <TextField id="customTxtField"/>
  <Label id="customLblMoveUp" visible="false"/>
  <Label id="customLblHintText"/>
</Alloy>
```

views/test.xml
```
<Alloy>
    <Window id="test" title="Test" onClose="cleanup">
      <Require src="/components/textfield" id="txtOne" customTop="80" customHintText="Test" onChange="filter" />
    </Window>
</Alloy>
```

controllers/test.js
```
(function() {
  $.txtOne.addEventListener("focus", function(e){
    Ti.API.error("onFocus");         
  });
})();

function cleanup() {
    $.destroy();
}
function filter(){
   Ti.API.error($txtOne.getInputText());
}
```

controllers/components/textfield.js
```
exports.getInputText = getInputText;
//exports.setInputText = addEventListener;

function getInputText(){
    return $.customTxtField.value;
}
/*
function addEventListener(event, callback){
  switch(event) {
    case "focus":
      $.customTxtField.addEventListener('focus', callback);
      break;
    default:

  }
}
*/
```

**Comment by m1ga** [comment]
If you look at the textfield.js code and change it to this:
```js
exports.getInputText = getInputText;
exports.addEventListener = addEventListener;  // <- was missing

function getInputText() {
	return $.customTxtField.value;
}

function addEventListener(event, callback){
  switch(event) {
    case "focus":
      $.customTxtField.addEventListener('focus', callback);
      break;
    default:
  }
}
```
it works fine. I see the `onFocus` error in the logs.

**Reply by beppo-ivel** [reply]
Thanks. Was missing this typo. Why there is no Alloy-Property compare to "onChange"?

**Reply by m1ga** [reply]
there is:

```xml
<Alloy>
    <Window>
        <TextField onChange="onChange"/>
    </Window>
</Alloy>
```

```js
function onChange(e){
	console.log(e.value);
}
$.index.open();
```

**Reply by beppo-ivel** [reply]
I mean why there is no Alloy property "onFocus" compare to the "onChange" property?

**Reply by m1ga** [reply]
again, there is: 

```xml
<Alloy>
    <Window>
        <TextField onFocus="onFocus"/>
    </Window>
</Alloy>
```

```js
function onFocus(e){
	console.log("focus");
}
$.index.open();
```

**Reply by beppo-ivel** [reply]
If I add the onFocus property in the composition from start post `...<Require src="/components/textfield" id="txtOne" customTop="80" customHintText="Test" onChange="filter" onFocus="onFocus" />...` , then the function `onFocus` is not executed. Only the pattern `exports.addEventListener = addEventListener; ` works as expected.

**Reply by m1ga** [reply]
`<Require>` is not the same as an element like `<TextField>`. Check the generated classic JS files to see what it generates. Can't tell you right away how it is applied but I would check `$.args` to see if the function is in there or use it with the `exports` part.

**Reply by m1ga** [reply]
ok quickly had a look
```js
$.__views["__alloyId1"].on('focus', onFocus)
```
it will add a "local" event that you can trigger yourself, not the event that is bound to the JAVA focus.

**Reply by beppo-ivel** [reply]
> Check the generated classic JS files to see what it generates.
You mean the `Resources/android/alloy/alloy.js`?

**Reply by m1ga** [reply]
I use `/build/android/assets/alloy/controllers/` or `/Resources/android/alloy/controllers/`. Think they should be the same but the first one is minified

**Reply by beppo-ivel** [reply]
> ok quickly had a look
> 
> ```js
> $.__views["__alloyId1"].on('focus', onFocus)
> ```
> 
> it will add a "local" event that you can trigger yourself, not the event that is bound to the JAVA focus.

Can you please describe it in more detail. There is a difference between onChange and onFocus property in `<Require src="/components/textfield" id="txtOne" customTop="80" customHintText="Test" onChange="filter" onFocus="filter" />` . How to trigger onFocus in this case, if I don't want to use the export part?

**Reply by m1ga** [reply]
`$.txtOne.trigger("focus")` should trigger that focus. It is not the native (e.g. Java) focus event but an event in JS (https://blog.axway.com/product-insights/discontinued/titanium/global-app-events-in-titanium-using-backbone-js)
So it doesn't trigger when the native textfield is trigger the focus event.
